src/net_processing.cpp

```diff
@@ -4303,6 +4324,8 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
     if (!peer) return false;
     const Consensus::Params& consensusParams = m_chainparams.GetConsensus();
 
+    LOCK(m_mutex_message_handling);
```

Comment:
>  /** Message handling mutex.
>    *  Message processing is single-threaded, so anything only accessed
>    *  by ProcessMessage(s) or SendMessages can be guarded by this mutex,
>    *  which guarantees it's only accessed by a single thread.

I'm confused. This mutex ends up protecting everything accessed in those functions. For example, in `SendMessages`, it also protects `MaybeDiscourageAndDisconnect` that accesses things like `peer.m_misbehavior_mutex`, which can be accessed by the validation thread. So this mutex is guarding more things than what is _only_ accessed by the `ProcessMessage(s)` and `SendMessages` thread.

I suppose it's safe that this mutex also ends up guarding things that are accessed by multiple threads, but I find the comment confusing.

Reply:
If you have:

```c++
Mutex mutex;
int x GUARDED_BY(mutex);
int y;
static void A() { LOCK(mutex); ++x; ++y; }
static void B() { LOCK(mutex); --x; }
void ThreadA() { for (int i = 0; i < 100; ++i) { A(); } }
void ThreadB() { for (int i = 0; i < 100; ++i) { B(); } }
```

then I'd say that only `x` is guarded by the mutex, even though `y` is only actually accessed while the mutex is held. If you later  introduce

```c++
static void C() { y *= 2; }
```

then that would still be safe/correct if `ThreadA` were to start calling `C()`, but `y` would no longer always be protected by the mutex; conversely if `ThreadB` were to start calling `C()`, then `y` would no longer be thread safe, but because it's not guarded by anything, the compiler would not catch that bug.

I guess I'd look at it more as mutexes protect **code** from being run simultaneously; it's the GUARDED_BY **annotations** that protect the variables. If the variables aren't annotated, they don't have any automatic protections, only manual protection by the programmer constantly being careful.

Reply:
> I guess I'd look at it more as mutexes protect __code__ from being run simultaneously

IMO mutexes protect variables, not code.

From https://en.wikipedia.org/wiki/Lock_(computer_science)
> ... a mechanism that enforces limits on access to a __resource__ ...
> ... each thread cooperates by acquiring the lock before accessing the __corresponding data__ ...

Reply:
Read your second quote carefully -- it's not the mutex/lock that protects the data, it's the threads that protect it by cooperatively acquiring the lock before accessing the data.

It's 99% true that the point of using a mutex is to protect data, but it's not the mutex itself that ensures the data is protected, it's how you use the mutex. It's like looking both ways before crossing the road -- if you don't do that, you'll probably walk into traffic; but looking alone isn't enough, you have to make sure you do it before crossing the road, and depending on what you see, delay crossing the road. The point of annotations here is to ensure at compile time that you're not metaphorically stepping into the road before looking, or literally not accessing variables that are only meant to be accessed by a particular thread from another thread.

(I think I've repeated this point enough now; if there's something not clear, feel free to ask, but I'm not seeing any point to going around in circles another time)